crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -217,139 +231,90 @@ where
-                            };
-                            let Some(mut contents) = maybe_contents.take() else {
-                                drop(maybe_contents);
-
-                                return PieceCacheState {
-                                    stored_pieces,
-                                    free_offsets,
-                                    backend: new_cache,
-                                };
-                            };
-
-                            stored_pieces.reserve(new_cache.max_num_elements() as usize);
-
-                            while let Some(maybe_element_details) = contents.next().await {
-                                let (offset, maybe_piece_index) = match maybe_element_details {
-                                    Ok(element_details) => element_details,
-                                    Err(error) => {
-                                        warn!(
-                                            %index,
-                                            %error,
-                                            "Failed to get cache contents element details"
-                                        );
-                                        break;
-                                    }
-                                };
-                                match maybe_piece_index {
-                                    Some(piece_index) => {
-                                        stored_pieces.insert(
-                                            RecordKey::from(piece_index.to_multihash()),
-                                            offset,
-                                        );
-                                    }
-                                    None => {
-                                        free_offsets.push_back(offset);
-                                    }
-                                }
-
-                                // Allow for task to be aborted
-                                yield_now().await;
-                            }
-
-                            drop(maybe_contents);
-                            drop(contents);
-
-                            PieceCacheState {
-                                stored_pieces,
-                                free_offsets,
-                                backend: new_cache,
-                            }
-                        },
-                        format!("piece-cache.{index}"),
-                    )
-                },
-            )
-            .collect::<Result<Vec<_>, _>>();
 
-        let caches_futures = match maybe_caches_futures {
-            Ok(caches_futures) => caches_futures,
-            Err(error) => {
-                error!(
-                    %error,
-                    "Failed to spawn piece cache reading thread"
-                );
-
-                return;
+        // Build cache state of all backends
+        for (index, new_cache) in new_piece_caches.into_iter().enumerate() {
```

Comment:
I understand that now there is a single data structure containing everything, but it would be really nice for initialization performance to remain high.

You have removed `run_future_in_dedicated_thread`, meaning all the caches will now be processed sequentially instead of concurrently, which will massively slow down cache initialization performance, especially for large farmers.